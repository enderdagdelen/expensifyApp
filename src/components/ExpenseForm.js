
import React from 'react';
import ReactDOM from 'react-dom';
import moment from 'moment';
//date picker 
import { SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';
import { isInclusivelyBeforeDay } from 'react-dates';
//date picker

const now = moment();
//console.log(now.format('MMM Do, YYYY'));
 
class ExpenseForm extends React.Component {
    constructor(props){
        super(props);

        this.state={
            description: props.expense ? props.expense.description :'',
            note: props.expense ? props.expense.note :'',
            amount: props.expense ? props.expense.amount.toString() :'',
            createdAt: props.expense ? moment(props.expense.createdAt) :moment(),
            calenderFocused:false,
            error:''
        }
    }


    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>{
            return {
                description:description
            }
        })
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(()=>{
            return {
                note:note
            }
        })
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>{
                return {
                    amount:amount
                }
            })
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(()=>{
                return {
                    createdAt:createdAt
                }
            })
        }
        
    }

    onFocusChange = ({focused}) => {
        this.setState(()=>{
            return {
                calenderFocused:focused
            }
        })
    }

    onSubmit = (e) =>{
        if(!this.state.description || !this.state.amount){
            this.setState(()=>{
                return {
                    error:"Please Provide Description and Amount"
                }
            })
        }else{
            this.setState(()=>{
                return {
                    error:''
                }
            })            
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount, 10),
                note:this.state.note, 
                createdAt:this.state.createdAt.valueOf()
            })
        }
        e.preventDefault();
    }

    render() {
        return (
            <div className="container">
                {this.state.error && <p>{this.state.error}</p>}
                <form className="form-group" onSubmit={this.onSubmit}>
                    <input className="form-control" type="text" placeholder="Description" autoFocus
                    value={this.state.description} 
                    onChange={this.onDescriptionChange}/>

                    <input 
                    className="form-control my-3" 
                    type="text" 
                    placeholder="Amount" 
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />

                    <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange = {this.onDateChange}
                    focused = {this.state.calenderFocused}
                    onFocusChange = {this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(day)=>false}
                    isOutsideRange={day =>
                        !day.isAfter(moment().subtract(1, "month")) ||
                        !isInclusivelyBeforeDay(day, moment())
                    }
                    />

                    <textarea 
                    className="form-control" 
                    placeholder="Add a note for your expenses"
                    value={this.props.note} 
                    onChange = {this.onNoteChange}></textarea>
                   
                    <button type='submit' className="btn btn-primary btn-block">Add Expense</button>
                </form>
            </div>
        )
    }

}

export default ExpenseForm;
