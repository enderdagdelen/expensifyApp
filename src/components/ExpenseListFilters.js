import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';
import { sortByDate, sortByAmount, setStartDate,setEndDate} from '../actions/filters';
import { DateRangePicker} from 'react-dates';


export class ExpenseListFilters extends React.Component{
    
    state={
        calenderFocused:null
    }

    onDatesChange = ({startDate,endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calenderFocused) => {
        this.setState(()=>{
            return{
                calenderFocused:calenderFocused
            }
        })
    }

    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e)=>{
        {e.target.value==='date' ? this.props.sortByDate():this.props.sortByAmount()}
    }

render(){

    return(

        <div>
            <input type='text' value={this.props.filters.text} onChange= {this.onTextChange} />

            <select onChange={this.onSortChange}>
                <option value="date" >Date</option>
                <option value="amount">Amount</option>
            </select>

            <DateRangePicker
                startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                focusedInput={this.state.calenderFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                numberOfMonths={1}
                isOutsideRange={()=>false}
                showClearDates={true}
            />
        </div>
    )
}

}

const mapPropsToState = (state) => {
return {
    filters:state.filters
}
}

//5 adet dispatch var yani 5 adet burada yeni dispatch tanımlamamız gerekiyor. 
const mapDispatchToProps = (dispatch) => {
    return{

        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))

    }

}

export default connect(mapPropsToState,mapDispatchToProps)(ExpenseListFilters);




/*

    DERS 126'DAN ÖNCEKİ KODUN HALİ. TEST İÇİN DEĞİŞTİRİLDİ

class ExpenseListFilters extends React.Component{
    
        state={
            calenderFocused:null
        }

        onDatesChange = ({startDate,endDate})=>{
            this.props.dispatch(setStartDate(startDate));
            this.props.dispatch(setEndDate(endDate))
        }

        onFocusChange = (calenderFocused) => {
            this.setState(()=>{
                return{
                    calenderFocused:calenderFocused
                }
            })
        }

    render(){

        return(

            <div>
                <input type='text' value={this.props.filters.text} onChange= {(e)=>{
                    this.props.dispatch(setTextFilter(e.target.value));
                }} />

                <select onChange={(e)=>{
                    {e.target.value==='date' ? this.props.dispatch(sortByDate()):this.props.dispatch(sortByAmount())}
                }}>
                    <option value="date" >Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calenderFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true}
                />
            </div>
        )
    }

}

const mapPropsToState = (state) => {
    return {
        filters:state.filters
    }
}

export default connect(mapPropsToState)(ExpenseListFilters);

*/