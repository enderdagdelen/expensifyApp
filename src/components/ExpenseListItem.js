import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment'
import numeral from 'numeral';


export const ExpensesListItem = ({id,dispatch,description,amount,createdAt}) => (
    
    <div>
        <h3>Description:{description}</h3>
        <p>Amount:{numeral(amount).format('$0,0.00')}</p>
        <p>{moment(createdAt).format('D MMMM YYYY')}</p>
        <Link to={`/edit/${id}`} ><h3>Edit {description}</h3></Link>

    </div>
) 

const mapStateToProps = (state) => {
    return {
        expenses:state.expenses
    }
}

export default ExpensesListItem;

// ÖNEMLİ NOT
/* _______1_________
connect ile store'a component'i store'a bağladığımızda connect, 
dispatch fonksiyonunu da sanki state'ten gelirmiş gibi bizim state'in içine eklenir. Bir çeşit state'ten gelen properties ve connect'ten gelen property ve fonksiyonlar birleşirler, aynı kümelerdeki birleşim gibi. Bu sayede biz de dispatch'i destructured nesne içerisine yazdıktan sonra ancak component içerisinde kullanabiliyoruz. 

_________2__________
Bu videoya kadar dispatch'in sadece store ve props'lar ile kullanılabileceğini sanıyordum. Ama bu şekilde yalnız da kullanılabiliyormuş.

*/