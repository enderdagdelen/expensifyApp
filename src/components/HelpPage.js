
import React from 'react';


const HelpPage = () => (
    <div>
        You are the HELP page!!!
    </div>
)

export default HelpPage;
