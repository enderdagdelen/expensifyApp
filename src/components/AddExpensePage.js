import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

//Burada export etme sebebimiz bu componenti test ederken redux'e bağlanmamış olmasını istediğimizden.
export class AddExpensePage extends React.Component{

    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/')
    }

    render(){
        return(
            <div>

                <h1>Add Expense Page</h1>

                <ExpenseForm 
                    onSubmit = {this.props.onSubmit}
                />

            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return{
        addExpense: (expense) => dispatch(addExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage)



/* test aşamasında alttaki kodu üsttteki koda çevirdik sebebi vb notlarda yazıyor. not:15
const AddExpensePage = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
        onSubmit = {(expense)=>{

            props.dispatch(addExpense(expense));
            props.history.push('/');
        }} 
        />
    </div>
)




export default connect()(AddExpensePage);

*/