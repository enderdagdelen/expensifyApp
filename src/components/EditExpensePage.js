
import React from 'react';
import {connect} from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import { Link } from 'react-router-dom';

export class EditExpensePage extends React.Component{
    onSubmit=(expense)=>{
        this.props.editExpense(this.props.expense.id,expense)
        this.props.history.push('/');

    }

    onRemove=()=>{
        this.props.removeExpense({id:this.props.expense.id});
        this.props.history.push('/');
    }

    render() {

        return (
            <div>
                This is EditExpense Page
                <ExpenseForm 
                expense = {this.props.expense}
                onSubmit={this.onSubmit}/>
                <button type="button" onClick={this.onRemove}>Remove</button>
            </div>
        )
    }
}
    

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense)=>{
            return expense.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch, props)  => {
    return{
        editExpense: (id, expense)=> dispatch(editExpense(id,expense)),
        removeExpense:(data) => dispatch(removeExpense(data))
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);





/* test zamanında ders 125'den önceki hali 
const EditExpensePage = (props) => {

    return (
        <div>
            This is EditExpense Page {props.match.params.id}
            <ExpenseForm 
            expense = {props.expense}
            onSubmit={(expense)=>{
                props.dispatch(editExpense(props.match.params.id,expense))
               
            }}/>
            <button type="button" onClick={()=>{
                props.dispatch(removeExpense({id:props.expense.id}));
                props.history.push('/');
            }}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense)=>{
            return expense.id === props.match.params.id
        })
    }
}


export default connect(mapStateToProps)(EditExpensePage);

*/
