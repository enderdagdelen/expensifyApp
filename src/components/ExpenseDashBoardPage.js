import React from 'react';
import ExpenseList from './ExpenseList.js';
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpensesSumarry'


const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)


export default ExpenseDashboardPage;
