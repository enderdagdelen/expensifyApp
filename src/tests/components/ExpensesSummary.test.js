import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseSummary} from '../../components/ExpensesSumarry'

test('should correctly render expenses summary with 1 expense',()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expenseTotal={235}/>)

    expect(wrapper).toMatchSnapshot()
})


test('should correctly render multiple expenses',()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={23} expenseTotal={235334}/>)

    expect(wrapper).toMatchSnapshot()
})