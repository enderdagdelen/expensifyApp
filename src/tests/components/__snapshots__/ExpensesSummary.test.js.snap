// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expenses summary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     expenses totalling $21.33
  </h1>
</div>
`;

exports[`should correctly render multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     expenses totalling $21.33
  </h1>
</div>
`;
